Add CI/CD Observability content (#1266)
elastic/observability-docs: docs/en/observability/ci-cd-observability.asciidoc

Select any of those errors to view the specific information. In this case, it's errors relating to
the CI agent that stopped.

[role="screenshot"]
.Jenkins pipeline build error in Elastic Observability
image::images/jenkins-pipeline-build.png[Jenkins pipeline builds]

The Errors overview screen provides a high-level view of the exceptions that CI builds catch.
Similar errors are grouped to quickly see which ones are affecting your services
and allow you to take action to rectify them.

[role="screenshot"]
.Jenkins jobs and pipelines errors in Elastic Observability
image::images/jenkins-pipeline-errors.png[Jenkins pipeline build errors]

[role="screenshot"]
.Concourse CI pipeline execution as a trace in Elastic Observability
image::images/concourse-ci-traces.png[Concourse CI traces view]

[discrete]
[[ci-cd-developers]]
== Observability for developers

Development teams need to continuously optimize their ever-changing CI/CD pipelines to improve
their reliability while chasing for faster pipelines. There are requirements to document CI/CD pipelines
with trace visualizations to show what’s happening and improve performance and reliability
(flaky tests and pipelines).

> Integrating with many popular CIDevOps tools like Maven or Ansible using OpenTelemetry, Elastic Observability

Reviewer: ```suggestion
Integrating with many popular CICD/DevOps tools like Maven or Ansible using OpenTelemetry, Elastic Observability
```

solves these problems by providing deep insights into the execution of CI/CD pipelines.

[discrete]
[[ci-cd-visibility]]
=== CI/CD pipeline visibility

The visualization of CI/CD pipelines as distributed traces in Elastic Observability provides
documentation and health indicators of all your pipelines.

The APM Service view in Elastic Observability provides a view of all your instrumented CI/CD
servers.

[role="screenshot"]
.Jenkins servers in Elastic Observability

> image::images/jenkins-servers.png[Jenkins servers view]

Reviewer: I'll update a new version of this view with a subset of servers

The Service page provides more granular insights into your CI/CD workflows by breaking down health
and performance metrics by pipeline. You can sort and filter the list to quickly surface which
pipelines are the slowest or experience the most errors. In the example below, we have sorted pipelines
by average build duration to show the slowest ones.

[role="screenshot"]
.A Jenkins server in Elastic Observability

> image::images/jenkins-server.png[Jenkins server view]

Reviewer: This image is a bit blurry

Author: Hi @v1v I used the image from the google doc. Are the fields cropped on that image too? Thanks.

Reviewer: I just updated it with a more recent version

Author: Thanks!

[discrete]
[[ci-cd-pipelines]]
=== Individual pipeline visibility

Once you’ve identified the pipeline you want to troubleshoot, you can drill down to get more detailed
information about its performance over time. The pipeline summary shows a breakdown of duration and
failure rates across the pipeline’s individual builds and jobs to spot slowdowns or failures.

[role="screenshot"]
.Performance overview of a Jenkins pipeline in Elastic Observability
image::images/jenkins-pipeline-overview.png[Jenkins pipeline overview]

The pipelines and traditional jobs are instrumented automatically. If you spot a slow or failing
build and need to understand what’s happening, you can drill into the trace view of the build to look

> for the high duration or errorful jobs. You can then drill into the details to understand the

Reviewer: NIT:
Beat with me, I'm not a native speaker and I don't know a synonym for `drill into` , but since there are two occurrences quite closed, if it's possible to use another synonym?

source of the error.

[role="screenshot"]
.A Jenkins pipeline build as a trace in Elastic Observability
image::images/jenkins-pipeline-trace.png[Trace of a Jenkins pipeline build]

View the details of the build captured as labels of the distributed trace.

[role="screenshot"]
.Contextual attributes of a Jenkins pipeline execution in Elastic Observability

> image::images/jenkins-pipeline-trace.png[Attributes of a Jenkins pipeline execution]

Reviewer: This image is pointing to the same image in L147

Reviewer: IIUC, this is the image with the attributes

```suggestion
image::images/jenkins-pipeline-context.png[Attributes of a Jenkins pipeline execution]
```

+
[role="screenshot"]
image::images/kibana-url.png[Define Kibana URL]
+
. Finally, there are additional settings to configure:

* Endpoint certificates to use in communications.
* The service name and service namespace sent to the OpenTelemetry service.
* Timeouts and batch process times.
* Any steps you may want to omit from the data you send.
+
WARNING: You can export the OpenTelemetry configuration as environment variables to use them with other tools like otel-cli,
Ansible Otel plugin, and so on. If you enable this option, consider that you can potentially expose the credentials in
the console output.

To learn more about the integration of Jenkins with Elastic Observability, see https://plugins.jenkins.io/opentelemetry/[OpenTelemetry].

[discrete]
[[ci-cd-jenkins-dashbaords]]
==== Install Jenkins dashboards in {kib}

There are out of the box {kib} dashboards that help visualize some metrics for the CI/CD platform.

Using the {kibana-ref}/dashboard-import-api.html[Import API] or the {kib} UI, you
can https://github.com/jenkinsci/opentelemetry-plugin/tree/master/docs/dashboards/elastic[install dashboards]
that are compatible with version 7.12 or higher.

> 

Reviewer: TBC, the section regarding the Kibana import and visualisation.

Reviewer: @cyrille-leclerc , can you clarify what do you want to add here?

I took some screenshots

1) Import UI
<img width="1279" alt="Screenshot 2021-11-18 at 12 41 03" src="https://user-images.githubusercontent.com/2871786/142417862-1a2dda08-ec0e-42b1-816f-c2c0e615a8e1.png">

<img width="311" alt="Screenshot 2021-11-18 at 12 41 34" src="https://user-images.githubusercontent.com/2871786/142417897-2167cf14-ff43-4836-b05a-f037e7b0ec18.png">

2) Dashboard ready to be used

![image](https://user-images.githubusercontent.com/2871786/142418070-c6e1902b-e95c-4047-9991-63af9ab4cfa3.png)

3) Dashboard UI
<img width="2552" alt="Screenshot 2021-11-18 at 12 44 44" src="https://user-images.githubusercontent.com/2871786/142417934-68e9330e-b5bd-4d67-ba47-1b7672502006.png">

Reviewer: See https://github.com/elastic/observability-docs/pull/1266/commits/9f99ebeebdf659e80ad9b34098cd9ebc3fb671c3

Reviewer 2: I love the guidance through the GUI.
It's important to indicate users that they have to select "Check for Existing Object / Overwrite".
I'm wondering how much we should talk about the dashboard import API. Maybe it's enough to focus on the GUI based. approach or to have a short mention after the GUI based solution saying "It's also possible to import dashboards using the 'Import API'".

[discrete]

> [[ci-cd-mavin]]

Reviewer 3: ```suggestion
[[ci-cd-maven]]
```

> === Mavin

Reviewer 3: ```suggestion
=== Maven
```

    </extensions>
  </build>
</project>
----

You can now trigger to send the Maven build reporting performance data to Elastic Observability by
passing the configuration details as environment variables:

[source,bash]
----
export OTEL_EXPORTER_OTLP_ENDPOINT="https://elastic-apm-server.example.com:8200"
export OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer an_apm_secret_token"

mvn verify
----

You can instrument Maven builds without modifying the pom.xml file using the Maven
command line argument “-Dmaven.ext.class.path=...”

[source,bash]
----
export OTEL_EXPORTER_OTLP_ENDPOINT="https://elastic-apm-server.example.com:8200"
export OTEL_EXPORTER_OTLP_HEADERS="Authorization=Bearer an_apm_secret_token"

mvn -Dmaven.ext.class.path=path/to/opentelemetry-maven-extension.jar verify
----

You can also trigger your Maven builds from the CI platform and visualize the end-to-end
pipeline execution in Elastic Observability, including the detailed steps of your CI
pipeline and the Maven build.

When invoking Maven builds with Jenkins, it’s unnecessary to use environment variables
to configure the Maven build (`OTEL_EXPORTER_OTLP_ENDPOINT…`), to rely on Jenkins’ capability

> to inject OpenTelemetry configuration as environment variables (see here).

Reviewer: Missing link in `see here`

as stated in the https://docs.ansible.com/ansible/latest/collections/community/general/opentelemetry_callback.html#examples[examples] section.

The context propagation from the Jenkins job or pipeline is passed to the Ansible run. Therefore,
everything that happens in the CI is also shown in the traces.

[role="screenshot"]
.Visibility into your Ansible playbooks
image::images/jenkins-ansible-pipeline.png[Ansible playbooks in Jenkins]

[discrete]
[[ci-cd-otel-cli]]
=== Otel cli

https://github.com/equinix-labs/otel-cli[otel-cli] is a command-line tool for sending OpenTelemetry
traces, which is useful if instrumenting your scripts explicitly when no other implicit
integration is in place.

Using the otel-cli wrapper, you can configure your build scripts implemented in shell, make, or
another scripting language. For example, instrumenting the Makefile below with otel-cli helps
visualize every command in each goal as spans.

[source,bash]
----
# see https://blog.container-solutions.com/tagging-docker-images-the-right-way

NAME   := acmecorp/foo
TAG    := $$(git log -1 --pretty=%!H(MISSING))
IMG    := ${NAME}:${TAG}
LATEST := ${NAME}:latest
 
build:

>   @otel-cli exec --name 'docker build' docker build -t ${IMG} .
>   @otel-cli exec --name 'docker tag' docker tag ${IMG} ${LATEST}

Reviewer: ```suggestion
  @otel-cli exec \
    --name 'docker build' \
    docker build -t ${IMG} .
  @otel-cli exec \
    --name 'docker tag' \
    docker tag ${IMG} ${LATEST}
```

 
push:

>   @otel-cli exec --name 'docker push' --attrs "http.url=https://docker.elastic.dev" docker push ${NAME}

Reviewer: ```suggestion
  @otel-cli exec \
    --name 'docker push' \
    --attrs "http.url=https://docker.elastic.dev" \
    docker push ${NAME}
```

 
login:

>   @otel-cli exec --name 'docker login' --attrs 'rpc.system=grpc' docker login -u ${DOCKER_USER} -p ${DOCKER_PASS}

Reviewer: ```suggestion
  @otel-cli exec \
    --name 'docker login' \
    --attrs 'rpc.system=grpc' \
    docker login -u ${DOCKER_USER} -p ${DOCKER_PASS}
```

to assess the nature and the impact of the outage.

In the following image, a Jenkins CI build failed, and its exceptions are reported as errors.
Select any of those errors to view the specific information. In this case, it's errors relating to
the CI agent that stopped.

[role="screenshot"]
.Jenkins pipeline build error in Elastic Observability
image::images/jenkins-pipeline-build.png[Jenkins pipeline builds]

The Errors overview screen provides a high-level view of the exceptions that CI builds catch.
Similar errors are grouped to quickly see which ones are affecting your services
and allow you to take action to rectify them.

[role="screenshot"]
.Jenkins jobs and pipelines errors in Elastic Observability
image::images/jenkins-pipeline-errors.png[Jenkins pipeline build errors]

[role="screenshot"]
.Concourse CI pipeline execution as a trace in Elastic Observability
image::images/concourse-ci-traces.png[Concourse CI traces view]

[discrete]
[[ci-cd-developers]]
== Observability for developers

Development teams need to continuously optimize their ever-changing CI/CD pipelines to improve

> their reliability while chasing for faster pipelines. Visualizations of pipelines as distributed

Reviewer 4: This is very awkwardly phrased. Perhaps the word `for` is unintentional?

Author: Yes, great spot. Thanks!

[discrete]
[[ci-cd-ansible]]
=== Ansible

The Ansible OpenTelemetry plugin integration provides visibility into all your
Ansible playbooks. The plugin generates traces for each run and performance metrics to help
you understand which Ansible tasks or roles are run the most, how often they
fail, and how long they take to complete.

You can configure your Ansible playbooks with the https://docs.ansible.com/ansible/latest/collections/community/general/opentelemetry_callback.html[Ansible OpenTelemetry callback plugin].
It’s required to install the Opentelemetry python libraries and configure the callback
as stated in the https://docs.ansible.com/ansible/latest/collections/community/general/opentelemetry_callback.html#examples[examples] section.

The context propagation from the Jenkins job or pipeline is passed to the Ansible run. Therefore,
everything that happens in the CI is also shown in the traces.

[role="screenshot"]
.Visibility into your Ansible playbooks
image::images/jenkins-ansible-pipeline.png[Ansible playbooks in Jenkins]

> 

Reviewer: Do you think we should add an entry regarding the service map for the services used by the ansible playbooks?

[This PR ](https://github.com/ansible-collections/community.general/pull/3558/) already supports it.

For instance, the ansible playbooks executed in one of Jenkins controllers produced this service map

![image](https://user-images.githubusercontent.com/2871786/143047967-aac6deaf-48f8-4431-8126-3639f458b291.png)

Thats something we get for free with the ansible opentelemetry callback plugin

Author: @cyrille-leclerc What do you think?^^

Reviewer 2: it's a great idea @v1v thanks!

Reviewer: See https://github.com/elastic/observability-docs/pull/1266/commits/7c1be8f15a08d0771cc221f88554feb54152c42d

Select any of those errors to view the specific information. In this case, it's errors relating to
the CI agent that stopped.

[role="screenshot"]
.Jenkins pipeline build error in Elastic Observability
image::images/jenkins-pipeline-build.png[Jenkins pipeline builds]

The Errors overview screen provides a high-level view of the exceptions that CI builds catch.
Similar errors are grouped to quickly see which ones are affecting your services
and allow you to take action to rectify them.

[role="screenshot"]
.Jenkins jobs and pipelines errors in Elastic Observability
image::images/jenkins-pipeline-errors.png[Jenkins pipeline build errors]

[role="screenshot"]
.Concourse CI pipeline execution as a trace in Elastic Observability
image::images/concourse-ci-traces.png[Concourse CI traces view]

[discrete]
[[ci-cd-developers]]
== Observability for developers

Development teams need to continuously optimize their ever-changing CI/CD pipelines to improve
their reliability while chasing for faster pipelines. There are requirements to document CI/CD pipelines
with trace visualizations to show what’s happening and improve performance and reliability
(flaky tests and pipelines).

> Integrating with many popular CICD/DevOps tools like Maven or Ansible using OpenTelemetry, Elastic Observability

Reviewer 4: ```suggestion
Integrating with many popular CI/CD and DevOps tools like Maven or Ansible using OpenTelemetry, Elastic Observability
```

* Endpoint certificates to use in communications.
* The service name and service namespace sent to the OpenTelemetry service.
* Timeouts and batch process times.
* Any steps you may want to omit from the data you send.
+
WARNING: You can export the OpenTelemetry configuration as environment variables to use them with other tools like otel-cli,
Ansible Otel plugin, and so on. If you enable this option, consider that you can potentially expose the credentials in
the console output.

To learn more about the integration of Jenkins with Elastic Observability, see https://plugins.jenkins.io/opentelemetry/[OpenTelemetry].

[discrete]
[[ci-cd-jenkins-dashbaords]]
==== Install Jenkins dashboards in {kib}

There are out of the box {kib} dashboards that help visualize some metrics for the CI/CD platform.

Using the {kibana-ref}/dashboard-import-api.html[Import API] or the {kib} UI, you
can https://github.com/jenkinsci/opentelemetry-plugin/tree/master/docs/dashboards/elastic[install dashboards]
that are compatible with version 7.12 or higher.

[discrete]
[[ci-cd-maven]]
=== Maven

The Maven OpenTelemetry plugin integration provides comprehensive visibility into all of your Maven builds.

> The plugin generates traces for each build and performance metrics to help you understand which maven goals

Reviewer 4: ```suggestion
The plugin generates traces for each build and performance metrics to help you understand which Maven goals
```

> or maven submodules are run the most, how often they fail, and how long they take to complete.

Reviewer 4: ```suggestion
or Maven submodules are run the most, how often they fail, and how long they take to complete.
```

The context propagation from CI pipelines (Jenkins job or pipeline) is passed to the Maven build
through the `TRACEPARENT` and `TRACESTATE` environment variables that match the https://www.w3.org/TR/trace-context/[W3C Trace Context specification].
Therefore, everything that happens in the CI platform is also shown in the traces.

You can configure your Maven project with the https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/maven-extension[Maven OpenTelemetry extension].

> For example, you can add the following snippet to your pom.xml file.

Reviewer 4: ```suggestion
For example, you can add the following snippet to your pom.xml file:
```

with trace visualizations to show what’s happening and improve performance and reliability
(flaky tests and pipelines).

> Integrating with many popular CICD/DevOps tools like Maven or Ansible using OpenTelemetry, Elastic Observability

Reviewer 2: This is the only place in the doc where we say CICD rather than CI/CD. Moreover, in the first sentence of the doc, we defined CI/CD="Continuous Integration and Continuous Delivery"
